### DISCUSSION TITLE
How to parse S3 Events and access fields

### DISCUSSION THREAD

**Discussion by i-mann** [awslabs/aws-sdk-rust#1147] [Q&A] [answered]
I'm trying to get the S3ObjectCreated field values out of an aws_sdk_s3::types::Event (via an SQS Message), and for the life of me I can't work out why.

I've read and reread the documentation around aws_sdk_s3::types::Event, and there's nothing that describes how that works - Event is only an enum, can only test for type, and there doesn't seem to be functions to get fields from the json-like body, it seems like the only way to extract these is using serde_json...

Is this correct? Have I missed something?

**Comment by aajtodd** [comment, marked as answer]
There is not a built in way to serialize/deserialize S3 event notifications.

You may be able to re-use [aws_lambda_events](https://docs.rs/aws_lambda_events/latest/aws_lambda_events/event/s3/index.html) which does provide serde implementations.

See also the S3 event message structure for reference: https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-content-structure.html

**Reply by i-mann** [reply]
Thanks for the fast answer! Is there any plans to include this in the near future?

**Reply by aajtodd** [reply]
Not currently no.

**Reply by i-mann** [reply]
To expand on this... something has changed in the May 17th release? How it's parsed? I recreated the json structure as structs and used serde to get the values I need, but then when reconnecting it up with the actual event again from AWS today... it's changed? Apart from seeing all the weird backslashes in the json that have now been removed.

There's no hint in the documents on what _should_ be done beyond that esoteric event match block, and what was working previously is below - now it fails at the `try_parse(body)` call, rather than being able to understand the event type.

It's a standard S3 Object Created event pushed over the event bridge to SQS...
```
    match aws_sdk_s3::types::Event::try_parse(body) {
        Ok(event) => {
            match event {
                Event::S3ObjectCreated => {
                      /// serde magic here
                },
                _ => // error code
            }
       Err(err) => // error code
     }
}
```
